### PR TITLE
EIP-712 signatures for login & signup

### LINKED ISSUE #181: Implement actual message signing for comments & auth
Our current auth and message posting has a sort of "fake" signing flow currently. These should be changed to actual EIP-712 signing. In addition to making all of the signing real, we'll also want to attach the signatures to the requests to the backend for validation.

This issue only covers the frontend, separate issues for comments and auth will be made.

### LINKED ISSUE #183: Validate auth signature
#181 is a prerequisite. When a user signs up or signs in, they need to sign a message to prove they own the address in question. `POST /users` and a newly needed `POST /users/login` endpoint will need to validate the signature, and respond if it's valid.

### PR DESCRIPTION
**Pull request by wbobeirne**
Closes #181, closes #183.

### What This Does

* Implements the full EIP-712 spec for message signing.
* Reworks signatures in redux / api to include raw data.
* Makes use of @dternyak's EIP-712 validating service to check them on the backend for POST /user signup and newly added POST /user/auth login.

### Steps to Test

1. Create a new account, confirm signature was required and it worked.
2. Sign out, then sign back in, confirm signature was required and it worked.
3. Refresh page, confirm you were logged back in.
4. Try canceling signature prompt for steps 1 & 2, confirm you got an error.
5. Load the page as one metamask account, and then switch to another.

I'm unsure as to how you could easily test an "invalid" signature, but I've manually edited the code to try it out and it seems to work.

## Screenshots

screen shot 2018-11-06 at 4 05 50 pm

### CODE REVIEW

File: frontend/client/api/axios.ts [hunk 1]
```diff
@@ -5,4 +5,23 @@ const instance = axios.create({
...
+      if (err.response.data.message) {
+        err.message = err.response.data.message;
+      }
+      // Some flask middlewares return error data like this
+      if (err.response.data.data) {
+        err.message = err.response.data.data;
+      }
+    }
+    return Promise.reject(err);
+  },
```
Comment (wbobeirne) [inline frontend/client/api/axios.ts]: Error messages will now be what the server sends back, not "Request failed with code 500" or whatever it was.

File: frontend/client/modules/auth/persistence.ts [hunk 2]
```diff
@@ -5,5 +5,5 @@ export const authPersistConfig: PersistConfig = {
   key: 'auth',
   storage,
   version: 1,
-  whitelist: ['token', 'tokenAddress'],
+  whitelist: ['authSignature', 'authSignatureAddress'],
```
Comment (wbobeirne) [inline frontend/client/modules/auth/persistence.ts]: I did not bump version on this due to the fact that this breaks nothing, and we won't have many / any users with persisted auth that will care anyhow. In the future, I would write a migration for this.

File: frontend/types/index.ts [hunk 3]
```diff
@@ -1,43 +1,8 @@
...
-  ProposalUpdates,
-};
+export * from './user';
+export * from './social';
+export * from './create';
+export * from './comment';
+export * from './milestone';
+export * from './update';
+export * from './proposal';
+export * from './api';
```
Comment (AMStrix) [inline frontend/types/index.ts]: Nice!

**Comment by AMStrix** [comment]
I'm gettings some CORs weirdness on Claim Identity -> SIGN ->
screen shot 2018-11-07 at 11 42 31 am

**Comment by wbobeirne** [comment]
Ah geez, looks like at some point I'd removed a trailing slash from the post request. Must have done it sub-consciously then not tested the flow again.

I've configured the backend to not be strict about trailing slashes. Should avoid this variety of problem in the future.

**Pull request merged**